### PR TITLE
Feature/election

### PR DESCRIPTION
**Pull request by yujuiting**
[Implement leader election](https://github.com/coreos/etcd/blob/master/clientv3/concurrency/election.go)

```ts
const client = new Etcd3();
const election = new Election(client, 'singleton');

async function task() {
  const rid = `${Math.random()}`;
  // process will hang here until elected
  await election.campaign(rid);
  // do task
}
```

### CODE REVIEW

**Comment by yujuiting** [comment]
I can't reproduce test fail in node v6 and etcd v3.1.5. Is there any idea?

**Comment by connor4312** [comment]
Thanks for the contribution! The failure is probably some race which happens more frequently in the environment Travis runs tests in. I'll give this a review and try to spot the race this evening.

File: src/election.ts [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+    if (result.length === 0) {
+      return;
+    }
+
+    const lastKey = result[0];
+    const watcher = await this.namespace.watch().key(lastKey).create();
+    const deleteOrError = new Promise(async (resolve, reject) => {
+      // waiting for deleting of that key
+      watcher.on('delete', resolve);
+      watcher.on('error', reject);
```
Comment (petermelias) [inline src/election.ts]: as a point of discussion (I recently had to implement this exact functionality as an extension to this lib for work last month) -- I elected to also listen to the `disconnected` event on the watcher so that the election can react to being disconnected for too long as at that point the application may want to take action given that the election cannot know for sure where it is in the line of succession until the network partition ends. The Golang implementation does not make any distinction between particular types of `WatchResponse` errors but treats them all the same. In any case I'm curious what people's thoughts are on this one?

**Comment by petermelias** [comment]
context: had to implement a leader election module for work on top this lib last month -- basically did the same thing and ported the Golang implementation but it seemed like there were some subtle differences in terms of "nodifying" the experience so I'm curious to see this evolve. We don't use TypeScript or else I'd have PR'd but either way I'm keen to see this lib support this OOB. So I'll happily share my experience with implementing it.

File: test/election.test.ts [hunk 2]
```diff
@@ -0,0 +1,108 @@
+import { expect } from 'chai';
```
Comment (petermelias) [inline test/election.test.ts]: tests do not have any cases that simulate network partitions yet -- I discovered some interesting edge cases when writing tests against this algorithm that introduce connection interrupts similar to the lease tests in this [lib's core](https://github.com/mixer/etcd3/blob/master/test/lease.test.ts#L67)

File: src/election.ts [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+    }
+
+    const r = await this.namespace
+      .if(this.leaseId, 'Create', '==', this._leaderRevision)
+      .then(this.namespace.put(this.leaseId).value(value).lease(this.leaseId))
+      .commit();
+
+    if (!r.succeeded) {
+      this._leaderKey = '';
+      throw Election.notLeaderError;
```
[the remark this answers is not in the document]
Comment (yujuiting) [reply src/election.ts]: I'm sorry about naming that flag `isLeader`. It's just a alternate of `e.leaderSession`.
That flag represent a status that this process has register as a candidate.
I think we should always catch when proclaiming and resign if fail. And this status will be updated.

**Comment by yujuiting** [comment]
I forgot to implement `observe`. Leader should observe for election and resign when new leader elected or suicide when disconnected. I will send commit after few days.

File: src/election.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (petermelias) [reply src/election.ts]: Okay if it's just naming confusion that makes more sense-- maybe `isCampaigning` or something would be a better name for that method

File: src/election.ts (hunk 1 above)
Comment (yujuiting) [reply src/election.ts]: `disconnected` should be listen, I miss this exception. But I prefer throw those errors to application layer and let users to decision how to react with. Election is a fundamental instruction do no more thing but election IMO.

Comment (petermelias) [reply src/election.ts]: agree that application should get to decide what to do. In my work implementation I propagate all failures back up the application -- including losing the lease, not just watcher failures. That said, I chose to call `resign` in my implementation if there is an unrecoverable watcher failure (or unrecoverable lease failure) because `resign` clears the internal state of the election client such that when/if the application decides to `campaign` again, it is a clean "do-over" from the failure state. If we don't call `resign`, then while calling `campaign` again from the application should basically just attempt to re-acquire the lease and then update election state if it succeeds, it felt cleaner to me that the since a lost watcher/lost lease essentially compromises the client's view of the line-of-succession, that's logically equivalent to a resignation as in either case the client can no longer know where it stands until it reconnects.

Therefore, if we assume that the lease is unrecoverable and the watchers are unrecoverable, when the partition ends we are just at the back of the line again either way. This may just be a personal preference for managing the internal state machine, but it felt cleaner and it didn't seem to introduce anything the application couldn't control.

To be clear, we would only `resign` internally (as the Golang impl does) on an _unrecoverable_ network failure (lost lease or permanent watcher error) not a transient failure like a `keepAliveFailure` or watcher `disconnect`.

Food for thought I guess...

File: src/election.ts [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+    }
+  }
+
+  public async resign() {
+    this.throwIfNotReady();
```
Comment (connor4312) [inline src/election.ts]: If not ready, you can just return; the the node can't be a leader if we aren't ready, so we're already in the desired state (same reason we return from isCampaigning below.

File: src/election.ts [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+  public async ready() {
+    const leaseId = await this.lease.grant();
+
+    if (!this.isReady) {
+      this._leaseId = leaseId;
+    }
+  }
+
+  public async campaign(value: string) {
+    this.throwIfNotReady();
```
Comment (connor4312) [inline src/election.ts]: How about awaiting the lease grant instead? I.e. why do callers necessarily need to know to call `await ready()`?

File: src/election.ts [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+    try {
+      await this.waitForElected();
+    } catch (error) {
+      this._isCampaigning = false;
+      throw error;
+    }
+  }
+
+  public async proclaim(value: any) {
+    this.throwIfNotReady();
```
Comment (connor4312) [inline src/election.ts]: once again can be omitted, if we aren't ready the isCampaigning check will fail.

File: src/election.ts [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+ * await election.ready()
+ *
+ * // process will hang here until elected
+ * await election.campaign(id)
+ */
+export class Election {
+  public static readonly prefix = 'election';
+  public static readonly notLeaderError = new Error('election: not leader');
+  public static readonly noLeaderError = new Error('election: no leader');
+  public static readonly notReadyError = new Error('election: no ready');
```
Comment (connor4312) [inline src/election.ts]: Rather than using sentinel errors like the Go client does, it'd be nice to make each of these subclasses of Error in `errors.ts` and instantiate them anew when thrown. Amongst other benefits, this make the stacktrace be set correctly; as-is the stacktrace will be set when the consumer imports the class, rather than when it's thrown!

File: src/election.ts (hunk 1 above)
Comment (connor4312) [reply src/election.ts]: You don't necessarily need to listen to `disconnected`, aside from handling timeouts, since when we reconnect we'll ask etcd to replay any changes that have happened since the time when we disconnected. So from an application pov it'll just be in the "waiting" state for longer than normal but otherwise should be entirely consistent. Being able to time out would be nice, maybe allow an optional `timeout` parameter to be passed in `campaign()`.

I agree that we should attempt to `resign` in all failure cases.

File: src/election.ts [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+    const deleteOrError = new Promise(async (resolve, reject) => {
+      // waiting for deleting of that key
+      watcher.on('delete', resolve);
+      watcher.on('error', reject);
+    });
+
+    try {
+      await deleteOrError;
+    } catch (error) {
+      throw error;
```
Comment (connor4312) [inline src/election.ts]: Redundant

File: src/election.ts [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+    this.throwIfNotReady();
+
+    if (!this.isCampaigning) {
+      return;
+    }
+
+    await this.namespace
+      .if(this.leaseId, 'Create', '==', this._leaderRevision)
+      .then(this.namespace.delete().key(this.leaseId))
+      .commit();
```
Comment (connor4312) [inline src/election.ts]: You might consider invalidating the lease if this fails. Something spooky might happen that causes this particular call to fail -- we can't control that, but what we can control is whether the client ceases its heartbeats to etcd which would cause the lease to expire eventually.

File: src/election.ts [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+
+    if (!result.succeeded) {
+      try {
+        const kv = result.responses[0].response_range.kvs[0];
+        this._leaderRevision = kv.create_revision;
+        if (kv.value.toString() !== value) {
+          await this.proclaim(value);
+        }
+      } catch (error) {
+        await this.resign();
```
Comment (connor4312) [inline src/election.ts]: We should catch and silence any errors that happen here (as the Go implementation does) which would be safe if you use the lease invalidation strategy described below.

**Comment by yujuiting** [comment]
Thanks for @petermelias and @connor4312 
I will start to work on it after tomorrow.

File: src/election.ts (hunk 10 above)
Comment (yujuiting) [reply src/election.ts]: It there any good ways let us tell user this campaign call has occurred error and should not keep going without throw error? Or we just return a new campaign call? I have no idea what case could lead to fail at this moment.

File: test/election.test.ts [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+    });
+
+    it('should wait for leader', async () => {
+      await election.resign();
+
+      let currentLeaderKey = '';
+      election.on('leader', leaderKey => currentLeaderKey = leaderKey);
+
+      // waiting for watcher created
+      await sleep(30);
```
Comment (yujuiting) [inline test/election.test.ts]: We will create a watcher to wait for new leader when no leader found. Seems if there is a new leader elected before watcher created, we will miss this leader. Is it possible?
If it is, maybe we can create watcher anyway before looking for current leader for avoid this case, although it will waste resource.

File: src/election.ts [hunk 12]
```diff
@@ -0,0 +1,275 @@
...
+
+    this._leaderKey = '';
+    this._leaderRevision = '';
+    this._isCampaigning = false;
+  }
+
+  public async getLeader() {
+    const result = await this.namespace.getAll().sort('Create', 'Ascend').keys();
+    if (result.length === 0) {
+      throw new EtcdNoLeaderError();
```
Comment (yujuiting) [inline src/election.ts]: How about return null or empty string instead of throw error?

**Comment by yujuiting** [comment]
@connor4312 @petermelias 
Hey guys, sorry for late! I has completed this work and appreciate for your feedback. Thx!

**Comment by connor4312** [comment]
Thanks for the followup! Sorry about lack of review previously, I've been on vacation for the past two weeks. I'll take a look at this this evening / over the weekend.

File: src/election.ts [hunk 13]
```diff
@@ -0,0 +1,282 @@
+import BigNumber from 'bignumber.js';
+import * as EventEmitter from 'events';
```
Comment (SimonSchick) [inline src/election.ts]: Nit: Should be ` { EventEmitter }`

File: src/election.ts [hunk 14]
```diff
@@ -0,0 +1,282 @@
...
+  private readonly namespace: Namespace;
+  private lease: Lease;
+
+  private leaseId = '';
+  private _leaderKey = '';
+  private _leaderRevision = '';
+  private _isCampaigning = false;
+  private _isObserving = false;
+
+  public get leaderKey(): string { return this._leaderKey; }
```
Comment (SimonSchick) [inline src/election.ts]: cc @connor4312 Do we use `getters` anywhere in this project? This doesn't really follow the rest of the project's style.

File: src/election.ts [hunk 15]
```diff
@@ -0,0 +1,282 @@
...
+  constructor(public readonly parent: Namespace,
+              public readonly name: string,
+              public readonly ttl: number = 60) {
+    super();
+    this.namespace = parent.namespace(this.getPrefix());
+    this.lease = this.namespace.lease(ttl);
+    this.lease.on('lost', () => this.onLeaseLost());
+  }
+
+  public on(event: 'leader', listener: (leaderKey: string) => void): this;
```
Comment (SimonSchick) [inline src/election.ts]: You should move these declarations into an interface with the same name as the class, that way you don't have to de-define the default ones.

File: src/election.ts [hunk 16]
```diff
@@ -0,0 +1,282 @@
...
+  public get isReady(): boolean { return this.leaseId.length > 0; }
+  public get isCampaigning(): boolean { return this._isCampaigning; }
+  public get isObserving(): boolean { return this._isObserving; }
+
+  constructor(public readonly parent: Namespace,
+              public readonly name: string,
+              public readonly ttl: number = 60) {
+    super();
+    this.namespace = parent.namespace(this.getPrefix());
+    this.lease = this.namespace.lease(ttl);
```
Comment (SimonSchick) [inline src/election.ts]: Does this kickoff async logic? I know this has been done elsewhere in the code base but constructors with side effects tend to be rather problematic.

File: src/election.ts [hunk 17]
```diff
@@ -0,0 +1,282 @@
...
+    this.lease.on('lost', () => this.onLeaseLost());
+  }
+
+  public on(event: 'leader', listener: (leaderKey: string) => void): this;
+  public on(event: 'error', listener: (error: any) => void): this;
+  public on(event: string|symbol, listener: Function): this;
+  public on(event: string|symbol, listener: Function): this {
+    super.on(event, listener);
+    if (this.shouldObserve(event)) {
```
Comment (SimonSchick) [inline src/election.ts]: You probably want to use https://nodejs.org/api/events.html#events_event_newlistener instead, less hacky.

File: src/election.ts [hunk 18]
```diff
@@ -0,0 +1,282 @@
...
+  /* istanbul ignore next */
+  public once(event: string|symbol, listener: Function): this {
+    super.once(event, listener);
+    if (this.shouldObserve(event)) {
+      this.tryObserve();
+    }
+    return this;
+  }
+
+  public async ready() {
```
Comment (SimonSchick) [inline src/election.ts]: Nit: Inconsistent return types.

File: src/election.ts [hunk 19]
```diff
@@ -0,0 +1,282 @@
...
+      this.tryObserve();
+    }
+  }
+
+  private tryObserve(): void {
+    this.observe().catch(error => this.emitOrThrowError(error));
+  }
+
+  private shouldObserve(event: string|symbol): boolean {
+    switch (event) {
```
Comment (SimonSchick) [inline src/election.ts]: `return event === 'leader'`?

File: src/election.ts [hunk 20]
```diff
@@ -0,0 +1,282 @@
...
+    }
+
+    // only keep watch if listened
+    if (this.listenerCount('leader') > 0) {
+      this.tryObserve();
+    }
+  }
+
+  private tryObserve(): void {
+    this.observe().catch(error => this.emitOrThrowError(error));
```
Comment (SimonSchick) [inline src/election.ts]: Wouldn't this always lead to uncaught rejections if no handler this attached?, all `tryObserve` calls should be awaited.

File: src/election.ts [hunk 21]
```diff
@@ -0,0 +1,282 @@
...
+
+  private shouldObserve(event: string|symbol): boolean {
+    switch (event) {
+      case 'leader': return true;
+      default: return false;
+    }
+  }
+
+  private emitOrThrowError(error: any): void {
+    if (this.listenerCount('error') > 0) {
```
Comment (SimonSchick) [inline src/election.ts]: Does this assume the error handler fixes the related problem? This seems weird to me.

File: src/election.ts [hunk 22]
```diff
@@ -0,0 +1,282 @@
...
+    this.lease.on('lost', () => this.onLeaseLost());
+    this.leaseId = '';
+  }
+}
+
+async function waitForDelete(namespace: Namespace, key: string) {
+  const watcher = await namespace.watch().key(key).create();
+  const deleteOrError = new Promise(async (resolve, reject) => {
+    // waiting for deleting of that key
+    watcher.on('delete', resolve);
```
Comment (SimonSchick) [inline src/election.ts]: Use `once` instead of `on.

File: src/election.ts [hunk 23]
```diff
@@ -0,0 +1,282 @@
...
+  private onLeaseLost() {
+    this.lease = this.namespace.lease(this.ttl);
+    this.lease.on('lost', () => this.onLeaseLost());
+    this.leaseId = '';
+  }
+}
+
+async function waitForDelete(namespace: Namespace, key: string) {
+  const watcher = await namespace.watch().key(key).create();
+  const deleteOrError = new Promise(async (resolve, reject) => {
```
Comment (SimonSchick) [inline src/election.ts]: the resolver doesn't need to be async.

File: src/election.ts [hunk 24]
```diff
@@ -0,0 +1,282 @@
...
+    await deleteOrError;
+  } finally {
+    await watcher.cancel();
+  }
+}
+
+async function waitForDeletes(namespace: Namespace, keys: string[]) {
+  if (keys.length === 0) {
+    return;
+  } else if (keys.length === 1) {
```
Comment (SimonSchick) [inline src/election.ts]: redundant else

File: src/election.ts [hunk 25]
```diff
@@ -0,0 +1,282 @@
...
+    return waitForDelete(namespace, keys[0]);
+  }
+
+  const tasks = keys.map(key => async () => {
```
Comment (SimonSchick) [inline src/election.ts]: Do these need to be strictly executed in order? Otherwise you could use a `for of` loop instead.

File: src/errors.ts [hunk 26]
```diff
@@ -179,6 +179,16 @@ export class EtcdInvalidAuthTokenError extends Error {}
  */
 export class EtcdPermissionDeniedError extends Error {}
 
+/**
+ * Etcd leader election has no leader
+ */
+export class EtcdNoLeaderError extends Error {}
```
Comment (SimonSchick) [inline src/errors.ts]: Maybe some default error message?

File: test/election.test.ts [hunk 27]
```diff
@@ -0,0 +1,187 @@
+import { expect } from 'chai';
+import * as sinon from 'sinon';
+
+import { Election,
+         Etcd3 } from '../src';
+import { getOptions,
+         tearDownTestClient } from './util';
+
+const sleep = (t: number) =>  new Promise(resolve => setTimeout(resolve, t));
```
Comment (SimonSchick) [inline test/election.test.ts]: Nit: Spacing.

File: src/election.ts (hunk 16 above)
Comment (yujuiting) [reply src/election.ts]: Or move lease initialization to `ready`?

File: src/election.ts (hunk 18 above)
Comment (yujuiting) [reply src/election.ts]: Do you mean user may expect `ready` return a boolean?

File: src/election.ts (hunk 20 above)
Comment (yujuiting) [reply src/election.ts]: Is it proper to `observe` again (next tick or a little delay) after caught an error always? I worry about any situation I didn't consider.

File: src/election.ts (hunk 21 above)
Comment (yujuiting) [reply src/election.ts]: You are right. The error thrown only when:
- `proclaim` or `waitingForElected` during  `campaign`.
- `proclaim`

And I aware that `emitOrThrowError` may break mechanism of `campaign`. Without throw, campaign will always passed when occurred errors if `error` event is listened.

I think [about failure of `campaign`](https://github.com/mixer/etcd3/pull/66/files#r179031641) and throw directly if proclaim fail.

File: src/election.ts (hunk 10 above)
Comment (yujuiting) [reply src/election.ts]: Maybe another way is to use a 'elected' and 'resigned' events.

File: src/election.ts (hunk 15 above)
Comment (yujuiting) [reply src/election.ts]: Is it fine to mute rule `interface-name`?

File: src/election.ts (hunk 17 above)
Comment (yujuiting) [reply src/election.ts]: Cool! It did not appear in declaration, thanks!.

File: src/errors.ts (hunk 26 above)
Comment (yujuiting) [reply src/errors.ts]: Repeat name of this error class? I don't know what can I talk about this error 😣

File: src/election.ts [hunk 28]
```diff
@@ -4,6 +4,17 @@ import { EtcdNoLeaderError, EtcdNotLeaderError } from './errors';
...
 
+export interface Election { // tslint:disable-line interface-name
+  on(event: 'leader', listener: (leaderKey: string) => void): this;
+  /**
+   * errors are fired when:
+   * - observe error
+   * - recreate lease fail after lease lost
+   */
+  on(event: 'error', listener: (error: any) => void): this;
+  on(event: string|symbol, listener: Function): this;
```
Comment (SimonSchick) [inline src/election.ts]: This shouldn't be needed.

File: src/election.ts [hunk 29]
```diff
@@ -4,6 +4,17 @@ import { EtcdNoLeaderError, EtcdNotLeaderError } from './errors';
 import { Lease } from './lease';
 import { Namespace } from './namespace';
 
+export interface Election { // tslint:disable-line interface-name
+  on(event: 'leader', listener: (leaderKey: string) => void): this;
```
Comment (SimonSchick) [inline src/election.ts]: Docs?

File: src/election.ts [hunk 30]
```diff
@@ -20,7 +31,7 @@ export class Election extends EventEmitter {
   public static readonly prefix = 'election';
 
   private readonly namespace: Namespace;
-  private lease: Lease;
+  private lease: Lease|null = null;
```
Comment (SimonSchick) [inline src/election.ts]: Nit: Spacing.

**Comment by taliesins** [comment]
@connor4312 when is this going to be merged in?

**Comment by taliesins** [comment]
I created a demo that uses the features from this branch: https://github.com/taliesins/example-etcd3-nodejs

**Comment by ma53** [comment]
@connor4312 @petermelias @SimonSchick Will this PR be merged?  Are there other plans in motion to implement support for leader elections?

**Comment by connor4312** [comment]
Hi everyone. Thank you for your work on this, this has been tremendously useful and will be a welcome addition to the library.

I apologize for the delay in merging this PR. The work done here has been great and is super solid. However, I delayed merging it as there have been a larger set of fundamental shifts I wanted to make, and I needed some time to learn and understand the implementation from the Go version. I've made these changes in <link here>.

> **tl;dr**: I changed many of the internals--though the algorithms are the mostly the same, they're now in streams. Take a look and let me know what you think.

Primarily, there were a few things I wanted out of this which we didn't have initially:

### Cancellation and Streams

The Go API uses contexts for cancellation. We don't have those, and Node's grpc doesn't provide a means for asynchronous cancellation. The Go API also uses channels to deliver a stream of values. This PR uses event emitters, but I believe this leads to a less-friendly API than Go does. RxJS streams, however, allow for cancellation and are effectively like more-advanced Go channels.

In the linked PR, the primary campaign method is `campaignAdvanced()` which emits and RxJS stream of candidate state changes. Using this, consumers can time-out, retry on errors, and do other advanced operations against the state. If this stream errors or is unsubscribed from before the candidate becomes a leader, they automatically resign.

For example, clients could have a timeout and retry errors like so:

```js
await candidate.campaign('candidate1')
  .pipe(retry(3), timeout(1000 * 60)
  .toPromise();
```

Because RxJS is alien to many, I preserved the base campaign method as well, and added a second optional timeout method.

```js
// Campaign until we're the leader, wait forever potentially:
await candidate.campaign('candidate1');

// Throw an error if we aren't the leader in 5000ms:
await candidate.campaign('candidate1', 5000);
```

In this implementation, we can do a good job of fairly gracefully cleaning up after ourselves and never getting into an orphaned state.

### State and Ordering

Something that the Go API doesn't have is thread-safety against arbitrary calls. I wanted to be able to `campaign()`, `proclaim()`, and `resign()` in any order without errors or inconsistencies.

This is actually a weakness of the Go API (which was mirrored in this PR). Their `campaign()` internally [sets](https://github.com/etcd-io/etcd/blob/0af16979f8a579a977bf422ed8674ee416e4b64b/clientv3/concurrency/election.go#L81) the leader state after making a call to etcd, but if you try to proclaim before that call comes back, you'll [get an error](https://github.com/etcd-io/etcd/blob/0af16979f8a579a977bf422ed8674ee416e4b64b/clientv3/concurrency/election.go#L111)--and there's no way to inspect the internal state and see what's happening.

Likewise, calling `resign()` from on routine while campaign is running may leave you in a faulted state. I wanted to avoid leaving footguns around here. Once you call `campaign()`, you may resign or proclaim new values at any time. You cannot proclaim a new value if campaign() hasn't been called, or resign() was called, and you cannot campaign() twice--you need to use campaign for that.

### Observing

Observing is likewise a stream. We can set up our watchers in that stream (the WatchBuilder now exposes an observe() method), aggregate candidate keys, and emit when there's a new leader. Once clients unsubscribe, the watcher is automatically destroyed.

**Comment by petermelias** [comment]
I understand the reasoning here. I commented on #102 as well but tl;dr; while I see the value of using Rx to mimic Go abstractions I wanted to point out that Rx is a heavyweight abstraction to bring to a project just for a single feature.

That said, it appears to be done so if you're comfortable with the functionality and interface it seems like this is probably mergeable?

**Comment by connor4312** [comment]
I've got this merged now. I kept the same general approach but moved the observer and "campaign" to their own, RAII-style classes, which I think works better for managing state. This is published in 1.1.0 and is documented here: https://microsoft.github.io/etcd3/classes/election.html
